Mathlib/Tactic/Linter/GlobalAttributeIn.lean

```diff
@@ -0,0 +1,130 @@
+This is **not** obvious from reading the code, and in fact happened twice during the port,
+hence, we lint against it.
+
+*Example*: before this was discovered, `Mathlib/Topology/Category/TopCat/Basic.lean`
+contained the following code:
+```
+attribute [instance] ConcreteCategory.instFunLike in
+instance (X Y : TopCat.{u}) : CoeFun (X ⟶ Y) fun _ => X → Y where
+  coe f := f
+```
+Despite the `in`, this makes `ConcreteCategory.instFunLike` a global instance.
+
+This seems to apply to all attributes. For example:
+```lean
+theorem what : False := sorry
+
+attribute [simp] what in
+#guard true
+
+-- the `simp` attribute persists
+example : False := by simp  -- `simp` finds `what`
+
+theorem who {x y : Nat} : x = y := sorry
+
+attribute [ext] who in
+#guard true
+
+-- the `ext` attribute persists
+example {x y : Nat} : x = y := by ext
+```
+Therefore, we lint against this patter on all instance.
+
+For *removing* attributes, the `in` works as expected.
+```lean
+/--
+error: failed to synthesize
+  Add Nat
+-/
+#guard_msgs in
+attribute [-instance] instAddNat in
+#synth Add Nat
+
+-- the `instance` persists
+/-- info: instAddNat -/
+#guard_msgs in
+#synth Add Nat
+
+@[simp]
+theorem what : False := sorry
+
+/-- error: simp made no progress -/
+#guard_msgs in
+attribute [-simp] what in
+example : False := by simp
+
+-- the `simp` attribute persists
+#guard_msgs in
+example : False := by simp
+```
+-/
+
+open Lean Elab Command
+
+namespace Mathlib.Linter
+
+/-- Lint on any occurrence of `attribute [...] name in` which is not `local` or `scoped`:
+these are a footgun, as the attribute is applied *globally* (despite the `in`). -/
+register_option linter.globalAttributeIn : Bool := {
+  defValue := true
+  descr := "enable the globalAttributeIn linter"
+}
+
+namespace globalAttributeInLinter
+
+/-- Gets the value of the `linter.globalAttributeIn` option. -/
+def getLinterGlobalAttributeIn (o : Options) : Bool :=
+  Linter.getLinterValue linter.globalAttributeIn o
+
+/--
+`getGlobalAttributesIn? cmd` assumes that `cmd` represents a `attribute [...] id in ...` command.
+If this is the case, then it returns `(id, #[non-local nor scoped attributes])`.
+Otherwise, it returns `default`.
+-/
+def getGlobalAttributesIn? : Syntax → Option (Ident × Array (TSyntax `attr))
+  | `(attribute [$x,*] $id in $_) =>
+    let xs := x.getElems.filterMap fun a => match a.raw with
+      | `(Parser.Command.eraseAttr| -$_) => none
+      | `(Parser.Term.attrInstance| local $_attr:attr) => none
+      | `(Parser.Term.attrInstance| scoped $_attr:attr) => none
+      | `(attr| $a) => some a
+    (id, xs)
+  | _ => default
+
+/-- The `nonlocalAttributeInLinter` linter flags any occurrence of `attribute [instance] name in`,
+including scoped or with instance priority: despite the `in`, these define *global* instances,
+which can be rather misleading.
+Instead, remove the `in` or make this a `local instance`.
```

Comment:
this docstring is still outdated

Reply:
Indeed, thanks for re-checking. Updated.